File: howl-application/src/main/java/io/quartic/howl/HowlResource.java
package io.quartic.howl;

import io.quartic.common.uid.UidGenerator;
import io.quartic.howl.api.HowlStorageId;
import io.quartic.howl.storage.InputStreamWithContentType;
import io.quartic.howl.storage.StorageBackend;
import org.apache.commons.io.IOUtils;

import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.GET;
import javax.ws.rs.NotFoundException;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.StreamingOutput;
import java.io.IOException;
import java.io.InputStream;
import java.util.Optional;

import static io.quartic.common.uid.UidUtilsKt.randomGenerator;
import static javax.ws.rs.core.HttpHeaders.CONTENT_TYPE;

@Path("/{namespace}")
public class HowlResource {
    private final StorageBackend storageBackend;
    private final UidGenerator<HowlStorageId> howlStorageIdGenerator = randomGenerator(HowlStorageId::new);

    public HowlResource(StorageBackend storageBackend) {
        this.storageBackend = storageBackend;
    }

    @POST
    @Path("/")
    @Produces(MediaType.APPLICATION_JSON)
    public HowlStorageId uploadFile(@PathParam("namespace") String namespace,
                                    @Context HttpServletRequest request) throws IOException {
        HowlStorageId howlStorageId = howlStorageIdGenerator.get();
        storageBackend.put(request.getContentType(), namespace, howlStorageId.getUid(), request.getInputStream());
        return howlStorageId;
    }

    @PUT
    @Path("/{fileName}")
    public void uploadFile(@PathParam("namespace") String namespace,
        @PathParam("fileName") String fileName,
        @Context HttpServletRequest request) throws IOException {

        storageBackend.put(request.getContentType(), namespace, fileName, request.getInputStream());
    }

    private Response handleDownload(String namespace, String fileName, Long version) throws IOException {
        Optional<InputStreamWithContentType> file = storageBackend.get(namespace, fileName, version);

        return file.map( f ->
            Response.ok()
                .header(CONTENT_TYPE, f.contentType())
                .entity(((StreamingOutput) output -> {
                    try (InputStream inputStream = f.inputStream()) {
                        IOUtils.copy(inputStream, output);
                    }
                }))
                .build())
                .orElseThrow(NotFoundException::new);
    }

    @GET
    @Path("/{fileName}")
    public Response downloadFile(@PathParam("namespace") String namespace,
                                 @PathParam("fileName") String fileName) throws IOException {
        return handleDownload(namespace, fileName, null);
    }

    @GET
    @Path("/{fileName}/{version}")
    public Response downloadFile(@PathParam("namespace") String namespace,
                                 @PathParam("fileName") String fileName,
                                 @PathParam("version") Long version) throws IOException {
        return handleDownload(namespace, fileName, version);
    }
}

File: howl-application/src/main/kotlin/io/quartic/howl/HowlResource.kt
package io.quartic.howl

import io.quartic.common.uid.UidGenerator
import io.quartic.common.uid.randomGenerator
import io.quartic.howl.api.HowlStorageId
import io.quartic.howl.storage.Storage
import io.quartic.howl.storage.StorageCoords
import org.apache.commons.io.IOUtils
import javax.servlet.http.HttpServletRequest
import javax.ws.rs.*
import javax.ws.rs.core.Context
import javax.ws.rs.core.HttpHeaders.CONTENT_TYPE
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response
import javax.ws.rs.core.StreamingOutput

// TODO - get rid of 2D variants, and move {identity-namespace} into Resource path
@Path("/{target-namespace}")
class HowlResource(
        private val storage: Storage,
        private val howlStorageIdGenerator: UidGenerator<HowlStorageId> = randomGenerator { HowlStorageId(it) }
) {
    @POST
    @Produces(MediaType.APPLICATION_JSON)
    fun uploadAnonymousFile(
            @PathParam("target-namespace") targetNamespace: String,
            @Context request: HttpServletRequest
    ) = uploadAnonymousFile(targetNamespace, targetNamespace, request)

    @POST
    @Path("/{identity-namespace}")
    @Produces(MediaType.APPLICATION_JSON)
    fun uploadAnonymousFile(
            @PathParam("target-namespace") targetNamespace: String,
            @PathParam("identity-namespace") identityNamespace: String,
            @Context request: HttpServletRequest
    ): HowlStorageId {
        val howlStorageId = howlStorageIdGenerator.get()
        uploadFileOrThrow(targetNamespace, identityNamespace, howlStorageId.uid, request)
        return howlStorageId
    }

    @PUT
    @Path("/{filename}")
    fun uploadFile(
            @PathParam("target-namespace") targetNamespace: String,
            @PathParam("filename") fileName: String,
            @Context request: HttpServletRequest
    ) = uploadFile(targetNamespace, targetNamespace, fileName, request)

    @PUT
    @Path("/{identity-namespace}/{filename}")
    fun uploadFile(
            @PathParam("target-namespace") targetNamespace: String,
            @PathParam("identity-namespace") identityNamespace: String,
            @PathParam("filename") fileName: String,
            @Context request: HttpServletRequest
    ) {
        uploadFileOrThrow(targetNamespace, identityNamespace, fileName, request)
    }

    private fun uploadFileOrThrow(
            targetNamespace: String,
            identityNamespace: String,
            fileName: String,
            request: HttpServletRequest
    ) = storage.putData(
            StorageCoords(targetNamespace, identityNamespace, fileName),
            request.contentType,
            request.inputStream
    ) ?: throw NotFoundException()

    @GET
    @Path("/{filename}")
    fun downloadFile(
            @PathParam("target-namespace") targetNamespace: String,
            @PathParam("filename") fileName: String
    ) = downloadFile(targetNamespace, targetNamespace, fileName)

    @GET
    @Path("/{identity-namespace}/{filename}")
    fun downloadFile(
            @PathParam("target-namespace") targetNamespace: String,
            @PathParam("identity-namespace") identityNamespace: String,
            @PathParam("filename") fileName: String
    ): Response {
        val coords = StorageCoords(targetNamespace, identityNamespace, fileName)
        val (contentType, inputStream) = storage.getData(coords, null) ?: throw NotFoundException()  // TODO: provide a useful message
        return Response.ok()
                .header(CONTENT_TYPE, contentType)
                .entity(StreamingOutput {
                    inputStream.use { inputStream -> IOUtils.copy(inputStream, it) }
                })
                .build()
    }

}
